Testing our apis (#11)
Ada-Developers-Academy/core-unit-2: testing-our-apis/dot-env-and-the-test-db.md

The `python-dotenv` package wants us to:

1. Create a file in the project root named `.env` exactly
1. List this file (`.env`) in the `.gitignore` file

### `.gitignore`

The `.gitignore` file is a file that Git, our version-control system, uses as a list of files and folders to _automatically never track_.

Every file and folder listed in a `.gitignore` will be _untracked_ by Git. This helps our workflow, as it gives us the confidence that we'll never accidentally commit it to Git history.

### Adding `.env` to `.gitignore`

We can list `.env` as another file to ignore.

In the case that we already have a full `.gitignore` file, we can confirm that `.env` is listed by searching the file. `.env` only needs to be listed once.

## Populate the Environment Variables

`.env` files are formatted in the following way:

- Each line is a different environment variable
- Environment variables are named, separated by a `=`, and then given a value (with no spaces or quotes)
  - It's conventional to name environment variables with CAPITAL_CASING, just like constant variables

Beyond those guidelines, we are free to choose the best variable names and values for our project.

> For our Hello Books API, we want to designate a path to a development database and a testing database with the variables named `SQLALCHEMY_DATABASE_URI` and `SQLALCHEMY_TEST_DATABASE_URI`.

Author: Gonna bring this into an issues for a future refactor! I like the suggestion, it just happens to be a lot to refactor in one go. Thanks!

> For our Hello Books API, we want to designate a path to a development database and a testing database with the variables named `SQLALCHEMY_DATABASE_URI` and `SQLALCHEMY_TEST_DATABASE_URI`.

Reviewer: Yeah, I definitely wouldn't change this right now.

```
SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/hello_books_development
SQLALCHEMY_TEST_DATABASE_URI=postgresql+psycopg2://postgres:postgres@localhost:5432/hello_books_test
```

Now, let's make the test database that we just named!

## Create the Test Database

Just like how we set up the development database, we want to set up our test database in `psql`.

```SQL
CREATE DATABASE hello_books_test;
```

Let's also create a `book` table in this database.

```SQL
\c hello_books_test
CREATE TABLE book (
  id INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
  title VARCHAR(200),
  description TEXT
);
```

## `create_app` Now Handles Test Mode

> Now that we've set up environment variables and the test database, let's change some configurations in our app to use this.
> 
> Based on the current state of this Flask app, let's look at a replacement for the `app/__init__.py` file, then go through the differences.

Reviewer: minor revision in first passage
I'd just say this is revised and call out what to watch for
minor changes to discussion reminder
```suggestion
Now that we've set up environment variables and the test database, let's change some configurations in our app to make use of them.

Let's look at a revised version of the `app/__init__.py` file that includes the changes needed to handle test mode. We'll examine the differences following the code.
```

```python
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_migrate import Migrate
from dotenv import load_dotenv
import os

db = SQLAlchemy()
migrate = Migrate()
load_dotenv()

def create_app(test_config=None):
    app = Flask(__name__)

>     if test_config is None:

Reviewer: I assume we'll want to make a change like this following the slack discussion?
```suggestion
    if not test_config:
```

        app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
        app.config['SQLALCHEMY_DATABASE_URI'] = os.environ.get(
            "SQLALCHEMY_DATABASE_URI")
    else:
        app.config["TESTING"] = True
        app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = False
        app.config["SQLALCHEMY_DATABASE_URI"] = os.environ.get(
            "SQLALCHEMY_TEST_DATABASE_URI")

    db.init_app(app)
    migrate.init_app(app, db)

    from app.models.book import Book

>     with app.app_context():
>         db.create_all()

Reviewer: this will get done in the test runner code
```suggestion
```

    from .routes import books_bp
    app.register_blueprint(books_bp)

    return app
```

> | <div style="min-width:250px;"> Piece of Code </div> | Notes                                                                                                                                                                                                                                                              |

Reviewer: improve the word wrapping
```suggestion
| <div style="min-width:310px;"> Piece of Code </div> | Notes                                                                                                                                                                                                                                                              |
```

| --------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `from dotenv import load_dotenv`                    | The `python-dotenv` package specifies to import the package like this                                                                                                                                                                                              |

> | `import os`                                         | This built-in package will allow us to read the environment variables set in `.env`                                                                                                                                                                                |

Reviewer: this is a module, not a package
I'd just mention that it reads environment variables in general (even those not set through .env)
```suggestion
| `import os`                                         | This built-in module provides a way to read environment variables                                                                                                                                                                                |
```

> | `load_dotenv()`                                     | The `python-dotenv` package specifies to call this method and load the variables before using them                                                                                                                                                                 |

Reviewer: here we can mention that this bridges between the .env file and the os module
add explanations for why we are adding the param to create_app
```suggestion
| `load_dotenv()`                                     | The `python-dotenv` package specifies to call this method, which loads the values from our `.env` file so that the `os` module is able to see them.                                                                                                                                                                  |
| `def create_app( ... ):` | Add a new keyword parameter to `create_app` that we can use to customize its behavior |
| `test_config=None` | We have called the new parameter `test_config` which should receive a dictionary of configuration settings. It has a default value of `None`, making the parameter optional. |
```

> | `if test_config is None:`                           | Create a conditional based on the keyword argument `test_config`. Whenever we call `create_app()`, from now on, we can optionally pass in a test config dictionary. If `test_config` is `None`, that means we are not trying to run the app in a test environment. |

Reviewer: update for code agreement
most of this is moved up to the suggested new descriptions of the change to create_app
```suggestion
| `if not test_config:`                           | Check the keyword argument `test_config`. When we call `create_app()`, if `test_config` is falsy (`None` or empty), that means we are not trying to run the app in a test environment. |
```

| `app.config["TESTING"] = True`                      | Turns testing mode on                                                                                                                                                                                                                                              |
| `os.environ.get( ... )`                             | This syntax gets an environment variable by the passed-in name                                                                                                                                                                                                     |
| `"SQLALCHEMY_TEST_DATABASE_URI"`                    | This is the exact name of the **testing** database environment variable we defined in `.env`                                                                                                                                                                       |

### !callout-secondary

## Practical Tip: Reading > Writing

> Understanding how to read this code is more valuable than learning how to write it. Realistically, on many projects, test environments are set up. However, it's valuable to read this code in order to:

Reviewer: untangle the line about setting up test environments: test environments get set up , and???
```suggestion
Understanding how to read this code is more valuable than learning how to write it. On many projects, we set up test environments by referring back to code from a previous project. But we should strive to be able to read this code in order to:
```

- Have a better understanding of how the app runs, which will lead to better debugging
- Get familiar with the process of setting it up, in anticipation of future projects

### !end-callout

## Verify We Didn't Break It: Manually Test Development

> We'll test our test environment after we set up our ability to run tests!
> 
> However, let's verify that we didn't break our development environment. Our development environment should not have changed at all!
> 
> Let's run our app in development mode with `(venv) $ flask run`, and make some requests in Postman. We can manually test all of our functionality to ensure we haven't lost anything.

Reviewer: some minor changes to the first 2 passages
slipped in a reference to regressions (introduced in a suggestion in the previous lesson)
```suggestion
We'll test our test environment after we set up tests in the next lessons!

For now, let's verify that we didn't break our development environment. These changes should not have affected our development environment at all!

Let's run our app in development mode with `(venv) $ flask run`, and make some requests in Postman. We can manually test all of our functionality to ensure we haven't introduced any _regressions_, broken scenarios which used to work.
```
